product-list-backend/routes/main.js

```diff
@@ -0,0 +1,132 @@
+const router = require("express").Router();
+const faker = require("faker");
+const { Review, Product } = require("../models/product");
+
+router.param("product", function (req, res, next, id) {
+  const query = Product.findOne({ _id: id });
+
+  query.exec((err, product) => {
+    if (err) return next(err);
+    if (!product) return next(new Error("Can't find product."));
+
+    req.product = product;
+    next();
+  });
+});
+
+router.param("review", function (req, res, next, id) {
+  const reviewQuery = Review.findOne({ _id: id });
+
+  reviewQuery.exec((err, review) => {
+    if (err) return next(err);
+    if (!review) return next(new Error("Can't find review."));
+
+    req.review = review;
+    next();
+  });
+});
+
+router.get("/generate-fake-data", (req, res, next) => {
+  for (let i = 0; i < 90; i++) {
+    let product = new Product();
+
+    product.category = faker.commerce.department();
+    product.name = faker.commerce.productName();
+    product.price = faker.commerce.price();
+    product.image = "https://via.placeholder.com/250?text=Product+Image";
+
+    product.save((err) => {
+      if (err) throw err;
+    });
+  }
+  res.end();
+});
+
+router.get("/products", (req, res, next) => {
+  const perPage = 9;
+
+  // return the first page by default
+  const page = req.query.page || 1;
+  const category = req.query.category || null;
+  const price = req.query.price || null;
+  const query = req.query.query || null;
+
+  let findObj = {};
+  if (category) {
+    findObj.category = category;
+  }
+  if (query) {
+    findObj.name = new RegExp(query, "i");
+  }
+  const sortString = price && price === "lowest" ? "asc" : "desc";
+  Product.find(findObj)
+    .skip(perPage * page - perPage)
+    .limit(perPage)
+    .sort({ price: sortString })
+    .exec((err, products) => res.send(products));
+});
+
+//Returns a specific product by its id
+router.get("/products/:product", (req, res, next) => {
+  res.send(req.product);
+  next();
+});
+
+//returns products with reviews - extra
+router.get("/reviews", (req, res, next) => {
+  Product.find({ reviews: { $exists: true, $not: { $size: 0 } } }).exec(
+    (err, reviews) => {
+      res.send(reviews);
+    }
+  );
+});
+
+//Returns ALL the reviews for a product, but limited to 4 at a time
+router.get("/products/:product/reviews", (req, res, next) => {
+  const reviewsPerPage = 4;
+  const reviews = req.product.reviews;
+  const page = req.query.page || 1;
+
+  res.send(reviews);
+});
+
+//Creates a new product in the database
+router.post("/products", (req, res, next) => {
+  let newProduct = new Product({
+    category: req.body.category,
+    name: req.body.name,
+    price: req.body.price,
+    image: req.body.image,
+    reviews: [],
+  });
+  newProduct.save();
```

Comment:
This wouldn't work because it would send before the save is done.

And why call `next` here?